flask_static_digest/__init__.py

```diff
@@ -61,26 +62,40 @@ def static_url_for(self, endpoint, **values):
         :param values: Arguments of the URL rule
         :return: Static file path.
         """
-        if not self.has_manifest:
-            if self.host_url:
-                return self._prepend_host_url(self.host_url,
-                                              values.get("filename"))
-            else:
-                return flask_url_for(endpoint, **values)
-
         new_filename = {}
         filename = values.get("filename")
 
+        # note: this is taken from flask's url_for
+        # ( resolves relative endpoints )
+        req_ctx = _request_ctx_stack.top
+        if req_ctx is not None:
+            blueprint_name = req_ctx.request.blueprint
+
+            # If the endpoint starts with "." and the request matches a
+            # blueprint, the endpoint is relative to the blueprint.
+            if endpoint[:1] == ".":
+                if blueprint_name is not None:
+                    endpoint = f"{blueprint_name}{endpoint}"
+                else:
+                    endpoint = endpoint[1:]
+        # endnote
+
         if filename:
             # If the manifest lookup fails then use the original filename
             # so that Flask doesn't throw a 500, but instead a proper 404.
             # The above only happens if your template has an invalid filename.
-            new_filename["filename"] = self.manifest.get(filename, filename)
+            manifest = self.manifests.get(endpoint)
+            if manifest:
+                new_filename["filename"] = manifest.get(filename, filename)
 
         merged_values = {**values, **new_filename}
 
-        if self.host_url:
-            return self._prepend_host_url(self.host_url,
-                                          merged_values.get("filename"))
-        else:
-            return flask_url_for(endpoint, **merged_values)
+        flask_url = flask_url_for(endpoint, **merged_values)
+
+        return self._prepend_host_url(flask_url)
```

Comment:
Previously we returned 2 different things based on if `if self.host_url` was set when folks wanted to prepend an optional CDN URL. Does this break that?

Reply:
Note: [4a5827e](https://github.com/nickjj/flask-static-digest/pull/33/commits/4a5827e6fc984a9230da71291e4e78629baaf008) I've just removed to `_prepend_host_url` to make this even simpler and hopefully clearer.

So the old implementation of `static_url_for` had 4 different returns
``` python
if not self.has_manifest:
    if self.host_url:
        return ....
    else:
        return ...

if self.host_url:
   return ...
else:
   return ...
```

Yes, at first glance, there's way less logic branches in my implementation but that doesn't mean less cases are being handled. Hear me out.

``` python
# We have a set of values we're going to supply to url_for.
# At this point, 'new_filename' may or may not have a replaced filename
# The 'if self.has_manifest' case was already handled beforehand
merged_values = {**values, **new_filename}

# flask_url_for will give us a relative path which we always want to have as a starting point
flask_url = flask_url_for(endpoint, **merged_values)

# Here, self.host_url may or may not be None.
# if 'self.host_url' happens to be None, urljoin will simply return 'flask_url', unchanged
# The 'if self.host_url' case is already built into 'urljoin'
return urljoin(self.host_url, flask_url)
```

Main point here, the 4 different returns weren't so different after all. They don't really change our course of action.
In every case, we want to `url_for` with the appropriate values.
I every case, it's safe to simply call urljoin and have it figure things out for us.

To answer your question specifically,  `self.host_url` is not broken and is still working as intended.


And here, the 'urljoin' implementation if you're curious:
https://github.com/python/cpython/blob/3.11/Lib/urllib/parse.py#L513

Reply:
Does this still apply with my other comment of https://github.com/nickjj/flask-static-digest/pull/33#discussion_r1133083834?

Reply:
Also in the original implementation this block of code existed at the top:

```py
        if not self.has_manifest:
            if self.host_url:
                return self._prepend_host_url(self.host_url,
                                              values.get("filename"))
            else:
                return flask_url_for(endpoint, **values)
```

This allowed us to exit early from the function when there's no digested files (no manifest).

I think this might introduce a performance regression because quite a bit more logic needs to be executed. Technically if there's no manifest then there's nothing to do since we can use Flask's `url_for` directly.

Reply:
Having less logic branches leads to better performance. The CPU wants to speed through things. The more predictable your line of execution is, the faster it will go. I'd suggest reading about branchless programming and its benefits.

Besides, your early return optimizes for people who are wrongly using `static_url_for` in contexts where there are no manifests, it slows down every other correct use of `static_url_for`.
Frankly your concerns for performance are misguided. There's no doubt in my mind that performance will be impacted, but it's going to be from the added functionality, not a missing early return.

Reply:
> Besides, your early return optimizes for people who are wrongly using static_url_for in contexts where there are no manifests, it slows down every other correct use of static_url_for

It's expected to _always_ use `static_url_for` when you're referencing any type of static file (css, js, images, etc.). The reason there is you can use it in development without digests or in production with (or without) digests without having to switch around your template tags for both cases.

> Frankly your concerns for performance are misguided. There's no doubt in my mind that performance will be impacted

Would you be open to benchmarking the original function and yours with and without digested files?

I'll admit I have not read the inner workings of how a CPU will process code at that level but I don't understand how executing all of the code below from your patch will end up being faster or equal vs the original function that exits early when there's no work to do by directly calling Flask's `url_for` and returning.

I do think when there's no digest the new version will be measurably slower. I really have no idea how much slower, it's possible it won't be impactful, but if we have the numbers in front of us with a repeatable test when we can see.

You may have multiple calls to `static_url_for` on a page (css + js + couple of images). Let's call it 5 calls on a page.

If your version takes on average 500 microseconds more than the original then we're talking 2.5ms. If your end to end server response time is 25ms, that's a 10% performance regression to render a response.

Reply:
Well if I'm going to take the time to bench it, I'm also going to take the time to optimize it and that's probably a full rewrite of `static_url_for` incoming, as a fair warning.

But just to be clear, my implementation WILL BE SLOWER, that's the cost of new features. I'll be happy if I can make it to parity, if not then no biggie.

Reply:
Maybe we can use Python's `timeit` module to get an idea on both implementations before rewriting?

If we determine it's only 50 microseconds slower then I don't think we need to rewrite it because a 1% regression for making the code more straight forward is worth it. Especially if that 25ms needs to travel 50ms over the internet and then there's 300ms spent painting a page in a browser. Suddenly it's a 375ms vs 250 microsecond difference in which case it's not worth the effort rewriting it.

Reply:
Yeah it's a 25 microseconds increase for me, from 13 to 37 or so. Profiling reveals that most of that time is spent resolving endpoints which are now dynamic and `url_for` being slower than what `_prepend_host_url` was doing.

Reply:
Would you mind sharing the code and commands you ran to profile it? I believe you but it would be nice to have a repeatable test. Maybe we can even commit it somewhere for future regression testing (I can do this work later if you want).

Reply:
Just ran through this essentially, gives an average over 10000 calls
```
def create_app():
    app = Flask(__name__)

    app.config.from_object("config.settings")

    static_digest.init_app(app)

    @app.route("/")
    def index():

        start_time = timeit.default_timer()
        ITERS = 10000
        for i in range(ITERS):
            static_digest.static_url_for('static', filename='css/app.css')
            static_digest.static_url_for('static', filename='js/app.css')

        print((timeit.default_timer() - start_time) / ITERS)

        return render_template("index.html")

    return app
```
Kinda messy to set up since you'd need 2 virtual envs to compare. An arbitrary threshold for regressions could be set, but results will vary greatly on different hardware. You'd need an actual test suite powered by tox or something, could always start there.

Reply:
Ok thanks. Let's keep your version.

Beyond that, there's a couple of change requests on renaming a couple of things and we're good to go to merge.

Reply:
I had removed my benchmark so I redid it this morning and got much faster results somehow
old implementation:  7.55μs
new implementation: 9.8μs
My benchmark must have had some overhead or something, couldn't tell ya.

Simplest way to go about it just something like this:
```python
def create_app():
    app = Flask(__name__)

    app.config.from_object("config.settings")

    app.register_blueprint(bp_about)
    app.register_blueprint(bp_contact)

    flask_static_digest.init_app(app)

    @app.route("/")
    def index():
        ITERS = 10000
        time = timeit.timeit("flask_static_digest.static_url_for('static', " +
                             "filename='css/app.css')",
                             globals=globals(),
                             number=ITERS)
        print(time/ITERS)

        return render_template("index.html")

    return app
``` 
Not really something the current CI would support. You'd need an actual test suite. Probably with tox on top to manage multiple virtualEnvs if you're going to compare. Anyway the margin for error in benchmarking is rather large and I don't recommend introducing luck elements to your CI, that's generally a bad time waiting to happen.

Reply:
Yep, I wouldn't block on CI based on results. I don't think this project really needs it, but if it did there would be benchmarks you could manually run on demand.

Thanks for setting all of this up.